### PR TITLE
gh-95451: Update docs for wasm32-emscripten and -wasi platforms

### PR DESCRIPTION
**Pull request by tiran**
* Issue: gh-95451

### CODE REVIEW

File: Doc/library/intro.rst [hunk 1]
```diff
@@ -64,3 +64,50 @@ Notes on availability
...
   newer and glibc 2.27 or newer.
+
+.. _wasm-availability:
+
+Notes on WebAssembly platforms
+==============================
+
+The WebAssembly platforms ``wasm32-emscripten`` and ``wasm32-wasi`` provide a
+subset of POSIX APIs. WASM runtimes and browser are sandboxed and have
+limited access to host and rest of the world. Any standard library module,
```
Comment (merwok) [inline Doc/library/intro.rst]:
```suggestion
limited access to the host and external resources. Any standard library module
```

File: Doc/library/intro.rst [hunk 2]
```diff
@@ -64,3 +64,50 @@ Notes on availability
...
   *Availability: Linux >= 3.17 with glibc >= 2.27* requires both Linux 3.17 or
   newer and glibc 2.27 or newer.
+
+.. _wasm-availability:
+
+Notes on WebAssembly platforms
+==============================
+
+The WebAssembly platforms ``wasm32-emscripten`` and ``wasm32-wasi`` provide a
+subset of POSIX APIs. WASM runtimes and browser are sandboxed and have
```
Comment (mdboom) [inline Doc/library/intro.rst]:
```suggestion
subset of POSIX APIs. WASM runtimes and browsers are sandboxed and have
```

File: Doc/library/intro.rst [hunk 3]
```diff
@@ -64,3 +64,50 @@ Notes on availability
...
+The properties and behavior of Python on WebAssembly platforms depend on the
+`Emscripten`_ or `WASI`_-SDK version, WASM runtimes (browser, NodeJS, `wasmtime`_),
+and Python build time flags. WebAssembly, Emscripten, and WASI are evolving.
+Some features like networking may be supported in the future.
+
+For Python in the browser, users should consider `Pyodide`_ or `PyScript`_.
+PyScript is built on top of Pyodide, which itself is built on top of
+CPython and Emscripten. Pyodide provides access to browsers' JavaScript and
+DOM APIs as well as limited networking capabilities with JavaScript's
+``XMLHttpRequest`` and ``Fetch`` API.
```
Comment (mdboom) [inline Doc/library/intro.rst]: Looking forward to the day when that's just a "browser support module" installable in upstream CPython...

File: Doc/library/intro.rst [hunk 4]
```diff
@@ -64,3 +64,50 @@ Notes on availability
...
+
+* Process-related APIs are not available or always fail with an error. That
+  includes APIs that spawn new processes (:func:`~os.fork`,
+  :func:`~os.execve`), wait for processes (:func:`~os.waitpid`), send signals
+  (:func:`~os.kill`), or otherwise interact with processes. The
+  :mod:`subprocess` is importable but does not work.
+
+* The :mod:`socket` module is available, but is dysfunctional.
+
+* Some functions are stubs that either don't do anything and always return some
```
Comment (mdboom) [inline Doc/library/intro.rst]:
```suggestion
* Some functions are stubs that either don't do anything or always return some
```

File: Doc/library/intro.rst [hunk 5]
```diff
@@ -64,3 +64,50 @@ Notes on availability
...
+that uses processes, threading, networking (:mod:`socket`), signals, or
+other forms of inter-process communication (IPC), is either not available
+or may not work as on other Unix-like systems. File I/O, file system, and Unix
+permission-related functions are restricted, too. Emscripten does not permit
+blocking operations like blocking network I/O and :func:`~time.sleep`.
+
+The properties and behavior of Python on WebAssembly platforms depend on the
+`Emscripten`_ or `WASI`_-SDK version, WASM runtimes (browser, NodeJS, `wasmtime`_),
+and Python build time flags. WebAssembly, Emscripten, and WASI are evolving.
+Some features like networking may be supported in the future.
```
Comment (mdboom) [inline Doc/library/intro.rst]: Given that Emscripten and WASI runtimes are moving targets, there could at least be unofficial builds of Python build against future versions of emscripten where some of the features marked as unavailable become available.  Do we want to somehow flag the difference between "explicitly disabled in Python" vs. "not known to work in current environments"?  *Maybe* we want to add to this paragraph:

"Python features documented as unsupported may in fact work depending on your environment."

But *maybe* that's too confusing for most users to understand, and something we can address in the future as these things mature.

**Review by mdboom — APPROVED** [review]

File: Doc/library/intro.rst [hunk 6]
```diff
@@ -64,3 +64,50 @@ Notes on availability
...
+
+.. _wasm-availability:
+
+Notes on WebAssembly platforms
+==============================
+
+The WebAssembly platforms ``wasm32-emscripten`` and ``wasm32-wasi`` provide a
+subset of POSIX APIs. WASM runtimes and browser are sandboxed and have
+limited access to the host and external resources. Any standard library module
+that uses processes, threading, networking (:mod:`socket`), signals, or
```
Comment (ezio-melotti) [inline Doc/library/intro.rst]: Is there a reason why only the `socket` module is specified and not other modules?

File: Doc/library/intro.rst [hunk 7]
```diff
@@ -64,3 +64,50 @@ Notes on availability
...
+other forms of inter-process communication (IPC), is either not available
+or may not work as on other Unix-like systems. File I/O, file system, and Unix
+permission-related functions are restricted, too. Emscripten does not permit
```
Comment (ezio-melotti) [inline Doc/library/intro.rst]: Possibly part of a separate PR, but a few glossary entries about WebAssembly, Emscripten, and related terms would be useful IMHO.

File: Doc/library/intro.rst [hunk 8]
```diff
@@ -64,3 +64,50 @@ Notes on availability
...
+blocking operations like blocking network I/O and :func:`~time.sleep`.
```
Comment (ezio-melotti) [inline Doc/library/intro.rst]: I would rephrase this to avoid the double "blocking".

File: Doc/library/intro.rst (hunk 3 above)
Comment (ezio-melotti) [inline Doc/library/intro.rst]:
```suggestion
``XMLHttpRequest`` and ``Fetch`` APIs.
```

File: Doc/library/intro.rst [hunk 9]
```diff
@@ -64,3 +64,50 @@ Notes on availability
...
+
+.. TODO: update with information from Tools/wasm/README.md
+
+* Process-related APIs are not available or always fail with an error. That
+  includes APIs that spawn new processes (:func:`~os.fork`,
+  :func:`~os.execve`), wait for processes (:func:`~os.waitpid`), send signals
+  (:func:`~os.kill`), or otherwise interact with processes. The
+  :mod:`subprocess` is importable but does not work.
+
+* The :mod:`socket` module is available, but is dysfunctional.
```
Comment (ezio-melotti) [inline Doc/library/intro.rst]: "Dysfunctional" is not very informative -- could you elaborate a bit?

File: Doc/library/intro.rst [hunk 10]
```diff
@@ -64,3 +64,50 @@ Notes on availability
...
+
+* Some functions are stubs that either don't do anything and always return some
+  hardcoded value.
```
Comment (ezio-melotti) [inline Doc/library/intro.rst]:
```suggestion
* Some functions are stubs that either don't do anything and always return
  hardcoded values.
```

File: Doc/library/os.rst [hunk 11]
```diff
@@ -878,6 +901,9 @@ as internal buffering of data.
 
    .. availability:: Unix.
 
+      The function is a limited on Emscripten and WASI, see
```
Comment (ezio-melotti) [inline Doc/library/os.rst]: Did you mean:
```suggestion
      The function is limited on Emscripten and WASI, see
```
or is there a missing word after "limited"?

There are several more occurrences of this.

File: Doc/library/intro.rst (hunk 5 above)
Comment (tiran) [reply Doc/library/intro.rst]: Very few features are explicitly disabled. The ``config.site`` files disable a few syscalls that fail tests or make no sense. For example I disabled the effective user/group id function family. Emscripten returns 0, which some code interprets as `CAP_SYS_ADMIN` privileges. The `config.site` files in `Tool/wasm` directory have comments for special cases.

The target audience for the section is "curious users without previous WASM experience". I want to give them *some* ideas what does not work without going into too many technical details.

File: Doc/library/intro.rst (hunk 3 above)
Comment (tiran) [reply Doc/library/intro.rst]: I'm lobbying for a high level HTTP API in Python core.

File: Doc/library/intro.rst (hunk 6 above)
Comment (tiran) [reply Doc/library/intro.rst]: Which other modules do you expect? All networking modules are based on the socket module.

File: Doc/library/os.rst (hunk 11 above)
Comment (tiran) [reply Doc/library/os.rst]: Fixed, thanks

File: Doc/library/intro.rst (hunk 6 above)
Comment (ezio-melotti) [reply Doc/library/intro.rst]: I just found it a bit odd that there are no mentions about specific modules for processes (e.g. `subprocess`, `multiprocessing`) or threading, and one single example for networking.

File: Doc/library/asynchat.rst [hunk 12]
```diff
@@ -34,6 +34,7 @@ Typically an :class:`asyncore.dispatcher` server channel generates new
 :class:`asynchat.async_chat` channel objects as it receives incoming
 connection requests.
 
+.. include:: ../includes/wasm-notavail.rst
```
Comment (CAM-Gerlach) [inline Doc/library/asynchat.rst:37]: This module is currently scheduled for removal in 3.12. Is the intention to backport this to <=3.11, or is there some other reason this is here?

File: Misc/NEWS.d/next/Documentation/2022-07-29-23-02-19.gh-issue-95451.-tgB93.rst [hunk 13]
```diff
@@ -0,0 +1,2 @@
+Update module documentation with availability information on WebAssembly
+platforms wasm32-emscripten and wasm32-wasi.
```
Comment (CAM-Gerlach) [inline Misc/NEWS.d/next/Documentation/2022-07-29-23-02-19.gh-issue-95451.-tgB93.rst]:
```suggestion
Update library documentation with
:ref:`availability information <wasm-availability>`
on WebAssembly platforms ``wasm32-emscripten`` and ``wasm32-wasi``.
```

* Be more specific about the type of documentation
* Link the main availability information
* Use literal syntax for the literal platform names

File: Doc/library/intro.rst [hunk 14]
```diff
@@ -64,3 +64,57 @@ Notes on availability
...
   *Availability: Linux >= 3.17 with glibc >= 2.27* requires both Linux 3.17 or
   newer and glibc 2.27 or newer.
+
+.. _wasm-availability:
+
+Notes on WebAssembly platforms
+==============================
+
+The `WebAssembly`_ platforms ``wasm32-emscripten`` (`Emscripten`_) and
+``wasm32-wasi`` `WASI`_ provide a subset of POSIX APIs. WebAssembly runtimes
```
Comment (CAM-Gerlach) [inline Doc/library/intro.rst]:
```suggestion
``wasm32-wasi`` (`WASI`_) provide a subset of POSIX APIs. WebAssembly runtimes
```

Be consistent with previous

File: Doc/library/intro.rst [hunk 15]
```diff
@@ -64,3 +64,57 @@ Notes on availability
...
+networking, signals, or other forms of inter-process communication (IPC), is
+either not available or may not work as on other Unix-like systems. File I/O,
+file system, and Unix permission-related functions are restricted, too.
+Emscripten does not permit blocking I/O or other blocking operations like
+:func:`~time.sleep`.
+
+The properties and behavior of Python on WebAssembly platforms depend on the
+`Emscripten`_-SDK or `WASI`_-SDK version, WASM runtimes (browser, NodeJS,
+`wasmtime`_), and Python build time flags. WebAssembly, Emscripten, and WASI
+are evolving and moving standards. Some features like networking may be
```
Comment (CAM-Gerlach) [inline Doc/library/intro.rst]:
```suggestion
are evolving standards; some features like networking may be
```

* Avoid redundant/verbose word ("moving" with "evolving")
* Join with semicolon to reduce choppiness and improve flow

File: Doc/library/intro.rst [hunk 16]
```diff
@@ -64,3 +64,57 @@ Notes on availability
...
+DOM APIs as well as limited networking capabilities with JavaScript's
+``XMLHttpRequest`` and ``Fetch`` APIs.
+
+* Process-related APIs are not available or always fail with an error. That
+  includes APIs that spawn new processes (:func:`~os.fork`,
+  :func:`~os.execve`), wait for processes (:func:`~os.waitpid`), send signals
+  (:func:`~os.kill`), or otherwise interact with processes. The
+  :mod:`subprocess` is importable but does not work.
+
+* The :mod:`socket` module is available, but is dysfunctional. On Emscripten,
```
Comment (CAM-Gerlach) [inline Doc/library/intro.rst]:
```suggestion
* The :mod:`socket` module is available,
but is limited and behaves differently from other platforms.
On Emscripten,
```

Be more specific and avoid charged and potentially confusing words like "dysfunctional"

File: Doc/library/intro.rst [hunk 17]
```diff
@@ -64,3 +64,57 @@ Notes on availability
...
+  sockets are always non-blocking and require additional JavaScript code
+  and helpers on the server to proxy TCP through WebSockets, see
```
Comment (CAM-Gerlach) [inline Doc/library/intro.rst]:
```suggestion
  and helpers on the server to proxy TCP through WebSockets; see
```

Fix comma splice

File: Doc/library/intro.rst [hunk 18]
```diff
@@ -64,3 +64,57 @@ Notes on availability
...
+  `Emscripten Networking <https://emscripten.org/docs/porting/networking.html>`_
+  for more information. WASI snapshot preview 1 only permits sockets from an
+  existing file descriptor.
+
+* Some functions are stubs that either don't do anything and always return
+  hardcoded values.
```
Comment (CAM-Gerlach) [inline Doc/library/intro.rst:109]: Did you mean to refer to, e.g., the ``os`` module here? Another specific module? Or "other modules" in general? Whatever is the case, it should be explicitly specified, as I was rather confused what this was supposed to say. Furthermore, if "other modules" in general (as currently implied), it should be the last bullet, since it is a catchall that covers everything. Also, did did you mean "either ... or", or just "and" (no "either")? Right now it reads "either ... and" which doesn't really make grammatical sense.

File: Doc/library/intro.rst [hunk 19]
```diff
@@ -64,3 +64,57 @@ Notes on availability
...
+
+* Functions related to file descriptors, file permissions, file ownership, and
+  links are limited and don't support some operations. For example WASI does
```
Comment (CAM-Gerlach) [inline Doc/library/intro.rst]:
```suggestion
  links are limited and don't support some operations. For example, WASI does
```

Add missing comma

File: Doc/library/intro.rst [hunk 20]
```diff
@@ -64,3 +64,57 @@ Notes on availability
...
+* Process-related APIs are not available or always fail with an error. That
+  includes APIs that spawn new processes (:func:`~os.fork`,
+  :func:`~os.execve`), wait for processes (:func:`~os.waitpid`), send signals
+  (:func:`~os.kill`), or otherwise interact with processes. The
+  :mod:`subprocess` is importable but does not work.
+
+* The :mod:`socket` module is available, but is dysfunctional. On Emscripten,
+  sockets are always non-blocking and require additional JavaScript code
+  and helpers on the server to proxy TCP through WebSockets, see
+  `Emscripten Networking <https://emscripten.org/docs/porting/networking.html>`_
```
Comment (CAM-Gerlach) [inline Doc/library/intro.rst]: Minor style nitpick, but is there a reason this link was added inline, but all the others were not?

File: Doc/library/signal.rst [hunk 21]
```diff
@@ -24,6 +24,9 @@ explicitly reset (Python emulates the BSD style interface regardless of the
 underlying implementation), with the exception of the handler for
 :const:`SIGCHLD`, which follows the underlying implementation.
 
+On WebAssembly platforms ``wasm32-emscripten`` and ``wasm32-wasi``, signals
+are emulated and therefore behave differently. Several functions and signals
+are not available on these platforms.
```
Comment (CAM-Gerlach) [inline Doc/library/signal.rst:29]: If there are only "several", could they be specifically noted in the `availability` notes in the relevant functions? Maybe I missed it, but I couldn't find any.

File: Doc/library/os.rst [hunk 22]
```diff
@@ -34,6 +34,14 @@ Notes on the availability of these functions:
 
 * On VxWorks, os.popen, os.fork, os.execv and os.spawn*p* are not supported.
 
+* On WebAssembly platforms ``wasm32-emscripten`` and ``wasm32-wasi``, large
+  parts of the :mod:`os` module are not available or behave differently. API
```
Comment (CAM-Gerlach) [inline Doc/library/os.rst:38]:
```suggestion
  parts of the :mod:`os` module are not available or behave differently. APIs
```

Fix grammar (plural agreement)

File: Doc/library/os.rst [hunk 23]
```diff
@@ -34,6 +34,14 @@ Notes on the availability of these functions:
...
+  related to processes (e.g. :func:`~os.fork`, :func:`~os.execve`), signals
+  (e.g. :func:`~os.kill`, :func:`~os.wait`), and resources
+  (e.g. :func:`~os.nice`) are not available. Others like :func:`~os.getuid`
+  and :func:`~os.getpid` are emulated or stubs.
```
Comment (CAM-Gerlach) [inline Doc/library/os.rst:42]:
```suggestion
  (e.g. :func:`~os.nice`) are not available. Others, like :func:`~os.getuid`
  and :func:`~os.getpid`, are emulated or stubs.
```

Set off with commas

File: Doc/library/os.rst [hunk 24]
```diff
@@ -357,6 +365,9 @@ process and user.
 
    .. availability:: Unix.
 
+      The function is a stub on Emscripten and WASI, see
```
Comment (CAM-Gerlach) [inline Doc/library/os.rst:368]:
```suggestion
      The function is a stub on Emscripten and WASI; see
```

Fix grammar (comma splice)

File: Doc/library/os.rst [hunk 25]
```diff
@@ -427,6 +438,8 @@ process and user.
 
    Return the current process id.
 
+   The function is a stub on Emscripten and WASI, see
```
Comment (CAM-Gerlach) [inline Doc/library/os.rst:441]:
```suggestion
   The function is a stub on Emscripten and WASI; see
```

File: Doc/library/os.rst [hunk 26]
```diff
@@ -498,14 +511,17 @@ process and user.
 
    .. availability:: Unix.
 
+      The function is a stub on Emscripten and WASI, see
```
Comment (CAM-Gerlach) [inline Doc/library/os.rst:514]:
```suggestion
      The function is a stub on Emscripten and WASI; see
```

Fix grammar (comma splice)

File: Doc/library/os.rst [hunk 27]
```diff
@@ -680,6 +696,9 @@ process and user.
 
    Set the current numeric umask and return the previous umask.
 
+   The function is a stub on Emscripten and WASI, see
```
Comment (CAM-Gerlach) [inline Doc/library/os.rst:699]:
```suggestion
   The function is a stub on Emscripten and WASI; see
```

Fix grammar (comma splice)

File: Doc/library/os.rst [hunk 28]
```diff
@@ -878,6 +901,9 @@ as internal buffering of data.
 
    .. availability:: Unix.
 
+      The function is limited on Emscripten and WASI, see
```
Comment (CAM-Gerlach) [inline Doc/library/os.rst:904]:
```suggestion
      The function is limited on Emscripten and WASI; see
```

Fix grammar error (comma splice)

File: Doc/library/os.rst [hunk 29]
```diff
@@ -890,6 +916,9 @@ as internal buffering of data.
...
```
Comment (CAM-Gerlach) [inline Doc/library/os.rst:919]:
```suggestion
      The function is limited on Emscripten and WASI; see
```

File: Doc/library/os.rst [hunk 30]
```diff
@@ -978,6 +1007,9 @@ as internal buffering of data.
...
```
Comment (CAM-Gerlach) [inline Doc/library/os.rst:1010]:
```suggestion
      The function is limited on Emscripten and WASI; see
```

File: Doc/library/os.rst [hunk 31]
```diff
@@ -1449,6 +1481,9 @@ or `the MSDN <https://msdn.microsoft.com/en-us/library/z0kc8e3z.aspx>`_ on Windo
...
```
Comment (CAM-Gerlach) [inline Doc/library/os.rst:1484]:
```suggestion
      The function is limited on Emscripten and WASI; see
```

File: Doc/library/os.rst [hunk 32]
```diff
@@ -1874,6 +1912,9 @@ features:
       read-only flag with it (via the ``stat.S_IWRITE`` and ``stat.S_IREAD``
       constants or a corresponding integer value).  All other bits are ignored.
 
+      The function is limited on Emscripten and WASI, see
```
Comment (CAM-Gerlach) [inline Doc/library/os.rst:1915]:
```suggestion
      The function is limited on Emscripten and WASI; see
```

File: Doc/library/os.rst [hunk 33]
```diff
@@ -1900,6 +1941,9 @@ features:
 
    .. availability:: Unix.
 
+      The function is limited on Emscripten and WASI, see
```
Comment (CAM-Gerlach) [inline Doc/library/os.rst:1944]:
```suggestion
      The function is limited on Emscripten and WASI; see
```

File: Doc/library/os.rst [hunk 34]
```diff
@@ -3059,6 +3103,9 @@ features:
 
    .. availability:: Unix, Windows.
 
+      The function is limited on Emscripten and WASI, see
```
Comment (CAM-Gerlach) [inline Doc/library/os.rst:3106]:
```suggestion
      The function is limited on Emscripten and WASI; see
```

File: Doc/library/intro.rst [hunk 35]
```diff
@@ -64,3 +64,57 @@ Notes on availability
   libc version, then both conditions must hold. For example a feature with note
   *Availability: Linux >= 3.17 with glibc >= 2.27* requires both Linux 3.17 or
   newer and glibc 2.27 or newer.
+
+.. _wasm-availability:
+
+Notes on WebAssembly platforms
+==============================
```
Comment (CAM-Gerlach) [inline Doc/library/intro.rst]:
```suggestion
WebAssembly platforms
---------------------
```

I'd suggest making this a subsection of the top-level "Notes on availability" section, since this is a subtopic of it (availability -> WASM availability), and it also makes it a part of the section users reach when they click the link in the ``availability`` directive, which is now heavily being used for WASM (since it links to the "Availability" section) rather than a separate parallel section. It also reduces redundancy in the title.

File: Doc/library/asyncore.rst [hunk 36]
```diff
@@ -28,6 +28,8 @@
 This module provides the basic infrastructure for writing asynchronous  socket
 service clients and servers.
 
+.. include:: ../includes/wasm-notavail.rst
```
Comment (CAM-Gerlach) [inline Doc/library/asyncore.rst:31]: Likewise, this module is scheduled for removal in 3.12

File: Doc/library/crypt.rst [hunk 37]
```diff
@@ -38,6 +38,8 @@ this module.
 
 .. availability:: Unix, not VxWorks.
 
+.. include:: ../includes/wasm-notavail.rst
+
```
Comment (CAM-Gerlach) [inline Doc/library/crypt.rst:42]: Not super important because this is a deprecated module with removal scheduled in 3.13, but this does lead to a duplicated `availability` directive, which is not ideal.

File: Doc/library/ensurepip.rst [hunk 38]
```diff
@@ -36,6 +36,7 @@ when creating a virtual environment) or after explicitly uninstalling
    :pep:`453`: Explicit bootstrapping of pip in Python installations
       The original rationale and specification for this module.
 
+.. include:: ../includes/wasm-notavail.rst
```
Comment (CAM-Gerlach) [inline Doc/library/ensurepip.rst:39]: I don't have a super strong feeling here, but maybe better to put the availability note right above the ``seealso`` instead of below? It feels a little unexpected to not have the "See Also" be the last element in the introduction, but maybe that's just me.

File: Doc/library/fcntl.rst [hunk 39]
```diff
@@ -18,6 +18,8 @@ interface to the :c:func:`fcntl` and :c:func:`ioctl` Unix routines.  For a
 complete description of these calls, see :manpage:`fcntl(2)` and
 :manpage:`ioctl(2)` Unix manual pages.
 
+.. include:: ../includes/wasm-notavail.rst
```
Comment (CAM-Gerlach) [inline Doc/library/fcntl.rst:21]: Wouldn't it be better to put this after the introduction prose and right above the "``versionchanged`` annotations, which is where it typically goes inside of functions and elsewhere? Its confusing to me to have it stuck right in the middle of the introduction paragraphs.

File: Doc/library/getpass.rst [hunk 40]
```diff
@@ -14,6 +14,7 @@
 
 The :mod:`getpass` module provides two functions:
```
Comment (CAM-Gerlach) [inline Doc/library/getpass.rst]: It doesn't really make much sense to have a colon here if the next items listed are not the referred-to functions. Instead, I suggest either rewriting this like, e.,g.

```suggestion
The :mod:`getpass` module provides functions to
get the current user's name and prompt them for a password.
```

***OR*** moving it below the ``include``:

```suggestion
```

(continued)

File: Doc/library/getpass.rst [hunk 41]
```diff
@@ -14,6 +14,7 @@
...
 
+.. include:: ../includes/wasm-notavail.rst
```
Comment (CAM-Gerlach) [inline Doc/library/getpass.rst:16]: (continued from previous)

```suggestion

The :mod:`getpass` module provides two functions:

```

File: Doc/library/ssl.rst [hunk 42]
```diff
@@ -65,6 +65,7 @@ by SSL sockets created through the :meth:`SSLContext.wrap_socket` method.
 
    Use of deprecated constants and functions result in deprecation warnings.
 
+.. include:: ../includes/wasm-notavail.rst
```
Comment (CAM-Gerlach) [inline Doc/library/ssl.rst]: Typically ``availability`` is listed above ``versionadded``/``versionchanged``, at least in the various existing examples I checked. It would be good to try to be consistent about where users can expect to find this information, especially in Reference-focused docs such as this.

File: Doc/library/os.rst [hunk 43]
```diff
@@ -853,6 +872,8 @@ as internal buffering of data.
    .. versionchanged:: 3.4
       The new file descriptor is now non-inheritable.
 
+   .. availability:: not WASI.
```
Comment (CAM-Gerlach) [inline Doc/library/os.rst]:
```suggestion
   .. availability:: not WASI.
   
   .. versionchanged:: 3.4
      The new file descriptor is now non-inheritable.
```

In almost all the existing usages in this module (and essentially all those I checked elsewhere) as well as most of those added here, ``availability`` is always listed above ``versionchanged``. Let's be consistent here, especially since this is Reference-focused documentation, as a clear and consistent structure helps readers find the key information they are looking for easily without having to dig around for it.

File: Doc/library/os.rst [hunk 44]
```diff
@@ -867,6 +888,8 @@ as internal buffering of data.
    .. versionchanged:: 3.7
       Return *fd2* on success. Previously, ``None`` was always returned.
 
+   .. availability:: not WASI.
```
Comment (CAM-Gerlach) [inline Doc/library/os.rst]: Likewise, this should be moved above the ``versionchanged`` messages.

**Review by CAM-Gerlach — COMMENTED** [review]
A few questions on better explaining confusing or potentially mistyped content, and a bunch of small suggestions mostly fixing grammar and organization issues (including a bunch of identical comma splices), improving phrasing, and a few reST deficiencies.

File: Doc/library/intro.rst (hunk 6 above)
Comment (tiran) [reply Doc/library/intro.rst]: Makes sense. I have removed the reference to the socket module from the sentence. Initially I wanted to include examples for each missing feature, but then it turned out to make the sense much harder to read.

File: Doc/library/asynchat.rst (hunk 12 above)
Comment (tiran) [reply Doc/library/asynchat.rst:37]: Yes, the ticket is for 3.11 and 3.12. I forgot to include the backports 3.11 label.

File: Doc/library/os.rst (hunk 43 above)
Comment (tiran) [reply Doc/library/os.rst]: I have moved the availability information above version changed.

File: Doc/library/getpass.rst (hunk 40 above)
Comment (tiran) [reply Doc/library/getpass.rst]: I moved the include above the sentence.

File: Doc/library/ssl.rst (hunk 42 above)
Comment (tiran) [reply Doc/library/ssl.rst]: I moved it below the notes. It seems more logical to me.

File: .gitignore [hunk 45]
```diff
@@ -18,6 +18,7 @@
 *.rej
 *.swp
 *~
+*.bak
```
Comment (merwok) [inline .gitignore]: I have memories of this being discussed before (or maybe it was about `*.rej`): backup files should show up in `git status` so that the developer can have a look at the new and backup files to see if something is missing, and if satisfied delete the backup.  Hiding them defeats their usefulness.  You could argue that it should be ignored to ensure `git add .` doesn’t add it to the next commit, but developers should not add or commit files blindly, and we do have code review to catch these.

```suggestion
```

File: Doc/library/asyncore.rst (hunk 36 above)
Comment (merwok) [reply Doc/library/asyncore.rst:31]: I think the notes added are useful for 3.11, or at least harmless!

File: Doc/library/fcntl.rst (hunk 39 above)
Comment (merwok) [reply Doc/library/fcntl.rst:21]: IMO it is fine to have a short paragraph to introduce the module, then the availability note.
Many other docs (see below) have that layout.

**Comment by tiran** [comment]
Thank you for your feedback and reviews!
